server/api/search.ts

```diff
@@ -3,45 +3,68 @@ import { serverQueryContent } from '#content/server'
 
 export default defineEventHandler(async (event) => {
     // Fetch all documents
-    let docs = await serverQueryContent(event).find()
-
-    docs = await Promise.all(
-        docs
-            .filter(
-                (doc) => {
-                    return doc?._extension === 'md' &&
-                        doc?._draft === false &&
-                        doc?.hidden === true &&
-                        !doc?._empty
-                }
-            )
-            .map(
-                async ({ _id: id, _path: path, _dir: dir, title = '', description = '', body = undefined}) => {
-
-                    return {
-                        id,
-                        path,
-                        dir,
-                        title,
-                        description,
-                        keywords: body?.toc?.links.map(link => link?.text),
-                        body: extractTextFromAst(body) || ''
-                    }
-                }
-            )
+    const docs = (await serverQueryContent(event).where({$and: [
+        {
+            _extension: { $eq: "md" },
+        },
+        {
+            _partial: { $eq: false },
+        },
+        {
+            _draft: { $ne: true },
+        },
+        {
+            hidden: { $ne: true },
+        }
+    ]}).find()).filter(
+        (doc) => {
+            // Remove empty docs
+            return doc.body?.children.length !== 0;
+        }
     )
 
-    return docs
+    // TODO: Allow body limit customisation
+    const bodyLimit = Math.min(500, 20000 / docs.length)
+
+    const data = docs.map(
+        ({ _id: id, _path: path, _dir: dir, title = '', description = '', body = undefined}) => {
+
+            return {
+                id,
+                path,
+                dir,
+                title,
+                description,
+                keywords: body?.toc?.links.map(link => link?.text),
+                body: extractTextFromAst(body, bodyLimit) || ''
+            }
+        }
+    )
+
+    // Cache search results across client instances
+    setResponseHeader(event,'cache-control','max-age=300, must-revalidate')
```

Comment:
knowing it's something that occurs only at build time. Is it really necessary to set this header ?

Reply:
While it is statically set at build time, clients don't know the contents doesn't change and make a request each time. The cache header informs clients. For example web browsers may use a cached copy instead of requesting the data each time until the max-age. 

The use-case is allow separate client instances to share the cache. For example if the user opens two browser tabs within the max-age, the resource is fetched only once.